flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/allocator/StateLocalitySlotAssigner.java

```diff
@@ -157,13 +165,21 @@ public Map<AllocationID, Integer> calculateScore(
                                     .getMaxParallelism(),
                             parallelism.get(evi.getJobVertexId()),
                             evi.getSubtaskIndex());
+            // Estimate state size per key group. For scoring, assume 1 if size estimate is 0 to
+            // accommodate for averaging non-zero states
+            Optional<Long> kgSizeMaybe =
+                    stateSizeEstimates.estimate(evi.getJobVertexId()).map(e -> Math.max(e, 1L));
+            if (!kgSizeMaybe.isPresent()) {
+                continue;
+            }
```

Comment:
This means it's not a stateful operator and couldn't make use of local recovery anyway?

Reply:
Yes (the operator didn't add **any** managed keyed state to the latest checkpoint).

Reply:
> the operator didn't add any managed keyed state to the latest checkpoint

How does that interact with incremental checkpoints? Let's say no input is coming from the sources (I don't know, all customers are sleeping?); do we then not consider this state at all?

Reply:
We still need to consider this state: if we place the task on a different TM then it will have to download all its SST files (or am I missing something?)

There are two methods for keyed state:
1. `handle.getStateSize()` returns the full state size
2. [`handle.getCheckpointedSize()`](https://github.com/apache/flink/blob/464ded1c2a0497255b70f711167c3b7ae52ea0f7/flink-runtime/src/main/java/org/apache/flink/runtime/state/CompositeStateHandle.java#L62) returns "incremental" state size

As per above, `getStateSize` is used.

Reply:
> if we place the task on a different TM then it will have to download all its SST files (or am I missing something?)

That part is clear.

if incremental state is also covered (even if an operator didn't add any state to it in the latest checkpoint), then we're good.